src/group_theory/pi_hom.lean

```diff
@@ -0,0 +1,325 @@
+/-
+Copyright (c) 2022 Jocchim Breitner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joachim Breitner
+-/
+import group_theory.order_of_element
+import data.finset.noncomm_prod
+import data.fintype.card
+/-!
+# Canonical homomorphism from a pi group
+
+This file defines the construction of the canoncial homomorphism from a product group.
+
+Given a family of morphisms `ϕ i : H i →* G` for each `i : I` where elements in the
+image of different morphism commute, we obtain a canoncial morphism
+`pi_hom.hom : (Π (i : I), H i) →* G` that coincides with `ϕ`
+
+## Main definitions
+
+* `pi_hom.hom : (Π (i : I), H i) →* G` is the main homomorphism
```

Comment:
This is analogous to [`linear_map.lsum`](https://leanprover-community.github.io/mathlib_docs/linear_algebra/pi.html#linear_map.lsum), right?

Reply:
It looks similar, yes

Reply:
Although we don't have an equivalence here due to the extra requirement that elements from different ranges commute.

Reply:
Maybe it would be better to put these definitions and lemmas into the `monoid_hom` namespace? But what would be a good name? 
Given that we have `[monoid_hom.coprod](https://leanprover-community.github.io/mathlib_docs/algebra/group/prod.html#monoid_hom.coprod)`, probably `monoid_hom.copi` is the right name?

Reply:
So how about

> * `monoid_hom.copi : (Π (i : I), N i) →* M` is the main homomorphism
> * `monoid_hom.copi_on (S : finset S) : (Π (i : I), N i) →* M` is the homomorphism restricted to the
>    set `S`, and mostly of internal interest in this file, to allow inductive proofs.
> * `subgroup.copi : (Π (i : I), H i) →* G` is the specialization to `H i : subgroup G` and
>    the subgroup embedding.

Reply:
I don't think any mathematician has the instinct that `copi` refers to such a thing. Maybe `comm_coprod` is a good name, as `pi_hom` is analogous to the "fac" field in [colimit](https://leanprover-community.github.io/mathlib_docs/category_theory/limits/is_limit.html#category_theory.limits.is_colimit) (coproduct in this case, except (the finite-supported) pi isn't exactly the coproduct (i.e. free group), but its quotient by the pairwise commutators).

Reply:
But wouldn't `comm_coprod` be the one about `×` (if we'd add that), not the one about pi?

Reply:
In category theory "coproduct" isn't limited to binary, and is called [binary_coproduct](https://leanprover-community.github.io/mathlib_docs/category_theory/limits/shapes/types.html#category_theory.limits.types.binary_coproduct_cocone) when it's binary. Using `coprod` here may conflict with binary coproduct, but agrees with the usage in [finset.prod](https://leanprover-community.github.io/mathlib_docs/algebra/big_operators/basic.html#finset.prod). You could also start a [naming contest](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/naming.20contest) to gather opinions.

Reply:
I don't care too strongly, my interest is mostly in reducing the size of the stack of unmerged PRs. So whatever someone with merge permissions (@eric-wieser ?) would be happy with is good for me :-)

Reply:
The naming contest at https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Name.20for.20map.20from.20pi.20group was a bit inconclusive.

I’ll update the PR to go with `monoid_hom.noncomm_pi_coprod`, as suggested by @eric-wieser : The `noncomm_` prefix has precedence in [noncomm_prod](https://leanprover-community.github.io/mathlib_docs/data/finset/noncomm_prod.html#multiset.noncomm_prod), the `coprod` comes from [`monoid_hom.coprod`](https://leanprover-community.github.io/mathlib_docs/algebra/group/prod.html#monoid_hom.coprod) and the `pi` because `noncomm_coprod` would be for the binary product.

Reply:
Is this better now?